### PR TITLE
Duplicate messages in bidirectional_bridge fix

### PR DESCRIPTION
**Pull request by ArkadiuszNiemiec**
Issue mentioned in #43 and #111, solution based on discussion in #43. Branched from ardent as I am not able to build master on my system.

Tested by using parameter_bridge and multiple ROS1 and ROS2 publishers and subscribers on the same topic.

Thank you @wjwwood for your help!

This PR is copy of broken PR #112. Sorry for the inconvenience.

### CODE REVIEW

**Comment by ArkadiuszNiemiec** [comment]
@wjwwood It now passes all tests.

File: include/ros1_bridge/factory.hpp [hunk 1]
```diff
@@ -145,10 +149,22 @@ class Factory : public FactoryInterface
...
+    rclcpp::PublisherBase::SharedPtr ros2_pub = nullptr)
   {
+    if (ros2_pub) {
+      bool result = false;
+      auto ret = rmw_compare_gids_equal(&msg_info.publisher_gid, &ros2_pub->get_gid(), &result);
+      if (ret == RMW_RET_OK) {
+        if (result) {  // message GID equals to bridge's ROS2 publisher GID
+          return;  // do not publish messages from bridge itself
+        }
+      }
```
Comment (wjwwood) [inline include/ros1_bridge/factory.hpp]: The else case should be handled here. If `rmw_compare_gids_equal` fails, then we should probably raise an exception or at least log an error message to the console using `RCLCPP_ERROR` logging macro.

Comment (wjwwood) [reply include/ros1_bridge/factory.hpp]: You can get the underlying error message with `rmw_get_error_string_safe()`, and after getting it you should clear the error state with `rmw_reset_error()`.

Comment (ArkadiuszNiemiec) [reply include/ros1_bridge/factory.hpp]: Ok, I have found similar code on different repo. Added.

**Comment by ArkadiuszNiemiec** [comment]
> I still haven't reproduced it myself, but I assume you have seen that it fixes your original issue?

Yes, I have tested it by starting two nodes publishing on one topic. After including the fix the duplicates went away. I have started ROS2 node on the same topic and it also bridged perfectly.

**Comment by wjwwood** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/4202/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/1259/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/3484/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/4319/)

**Comment by wjwwood** [comment]
@ArkadiuszNiemiec looks like the linters are still unhappy about some trailing whitespace or something. Can you check `ament_uncrustify` and `ament_cpplint` one more time?

**Comment by ArkadiuszNiemiec** [comment]
@wjwwood Fixed, empty line was not empty :smile:

**Comment by wjwwood** [comment]
Thanks @ArkadiuszNiemiec!

**Pull request merged**